airbyte-integrations/connectors/source-recurly/source_recurly/source.py

```diff
@@ -2,71 +2,63 @@
 
 from airbyte_cdk.logger import AirbyteLogger
-from airbyte_cdk.models.airbyte_protocol import (
-    AirbyteCatalog,
-    AirbyteConnectionStatus,
-    AirbyteMessage,
-    AirbyteRecordMessage,
-    ConfiguredAirbyteCatalog,
-    Status,
-    Type,
+from airbyte_cdk.sources import AbstractSource
+from airbyte_cdk.sources.streams import Stream
+from recurly import ApiError, Client
+
+from .streams import (
+    RecurlyAccountCouponRedemptionsStream,
+    RecurlyAccountsStream,
+    RecurlyCouponsStream,
+    RecurlyExportDatesStream,
+    RecurlyInvoicesStream,
+    RecurlyMeasuredUnitsStream,
+    RecurlyPlansStream,
+    RecurlySubscriptionsStream,
+    RecurlyTransactionsStream,
 )
-from airbyte_cdk.sources.source import Source
-
-from .client import Client
 
 
-class SourceRecurly(Source):
+class SourceRecurly(AbstractSource):
     """
-    Recurly API Reference: https://developers.recurly.com/api/v2019-10-10/index.html
+    Recurly API Reference: https://developers.recurly.com/api/v2021-02-25/
     """
 
     def __init__(self):
-        super().__init__()
-
-    def check(self, logger: AirbyteLogger, config: json) -> AirbyteConnectionStatus:
-        client = self._client(config)
-        alive, error = client.health_check()
-        if not alive:
-            return AirbyteConnectionStatus(status=Status.FAILED, message=f"{error}")
-
-        return AirbyteConnectionStatus(status=Status.SUCCEEDED)
-
-    def discover(self, logger: AirbyteLogger, config: json) -> AirbyteCatalog:
-        client = self._client(config)
-
-        return AirbyteCatalog(streams=client.get_streams())
-
-    def read(
-        self, logger: AirbyteLogger, config: json, catalog: ConfiguredAirbyteCatalog, state: Dict[str, any]
-    ) -> Generator[AirbyteMessage, None, None]:
-        client = self._client(config)
-
-        logger.info("Starting syncing recurly")
-        for configured_stream in catalog.streams:
-            # TODO handle incremental syncs
-            stream = configured_stream.stream
-            if stream.name not in client.ENTITIES:
-                logger.warn(f"Stream '{stream}' not found in the recognized entities")
-                continue
-            for record in self._read_record(client=client, stream=stream.name):
-                yield AirbyteMessage(type=Type.RECORD, record=record)
-
-        logger.info("Finished syncing recurly")
-
-    def _client(self, config: json):
-        client = Client(api_key=config["api_key"])
-
-        return client
-
-    @staticmethod
-    def _read_record(client: Client, stream: str):
-        for record in client.get_entities(stream):
-            now = int(datetime.now().timestamp()) * 1000
-            yield AirbyteRecordMessage(stream=stream, data=record, emitted_at=now)
+        super(SourceRecurly, self).__init__()
+
+        self.__client = None
+
+    def check_connection(self, logger: AirbyteLogger, config: Mapping[str, Any]) -> Tuple[bool, Optional[Any]]:
```

Comment:
This class looks completely different, I'm assuming that's due to the significant framework changes since the original was written?

Reply:
Yes, the Recurly source was rewritten using CDK in a previous PR (it was using a legacy `base_python` code). This time I did another round of refactoring and inherited from some CDK base classes and implemented the abstract methods.